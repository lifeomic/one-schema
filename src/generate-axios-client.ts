import { OneSchemaDefinition } from '.';
import { generateEndpointTypes } from './generate-endpoints';
import { validateSchema } from './meta-schema';

export type GenerateAxiosClientInput = {
  spec: OneSchemaDefinition;
  outputClass: string;
};

export type GenerateAxiosClientOutput = {
  javascript: string;
  declaration: string;
};

export const generateAxiosClient = async ({
  spec,
  outputClass,
}: GenerateAxiosClientInput): Promise<GenerateAxiosClientOutput> => {
  validateSchema(spec);

  const declaration = [
    '/* eslint-disable */',
    "import { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';",
    '',
    await generateEndpointTypes(spec),
    `
export declare class ${outputClass} {

  constructor(client: AxiosInstance);

  ${Object.entries(spec.Endpoints)
    .map(([endpoint, { Name }]) => {
      const [method] = endpoint.split(' ');
      const paramsName = method === 'GET' ? 'params' : 'data';

      return `
        ${Name}(
          ${paramsName}: Endpoints['${endpoint}']['Request'] &
            Endpoints['${endpoint}']['PathParams'],
          config?: AxiosRequestConfig
        ): Promise<AxiosResponse<Endpoints['${endpoint}']['Response']>>`;
    })
    .join('\n\n')}

  paginate<T extends { nextPageToken?: string; pageSize?: string }, Item>(
    request: (
      data: T,
      config?: AxiosRequestConfig
    ) => Promise<
      AxiosResponse<{ items: Item[]; links: { self: string; next?: string } }>
    >,
    data: T,
    config?: AxiosRequestConfig
  ): Promise<Item[]>;
}`.trim(),
  ].join('\n');

  const javascript = `
/* eslint-disable */

const substituteParams = (url, params) =>
  Object.entries(params).reduce(
    (url, [name, value]) => url.replace(":" + name, encodeURIComponent(value)),
    url
  );

const removePathParams = (url, params, encode) => 
  Object.entries(params)
    .filter(([key, value]) => value !== undefined)
    .reduce(
      (accum, [name, value]) =>
        url.includes(':' + name)
          ? accum
          : { ...accum, [name]: encode ? encodeURIComponent(value) : value },
      {}
    );

const parseQueryParamsFromPagingLink = (link) => {
  const params = new URLSearchParams(link.split('?')[1]);

  return {
    nextPageToken: params.get('nextPageToken'),
    pageSize: params.get('pageSize')
  };
};

class ${outputClass} {

  constructor(client) {
    this.client = client;
  }

  ${Object.entries(spec.Endpoints)
    .map(([endpoint, { Name }]) => {
      const [method, url] = endpoint.split(' ');
      const useQueryParams = ['GET', 'DELETE'].includes(method);
      return `
        ${Name}(data, config) {
          return this.client.request({
            ...config,
            method: '${method}',
            ${
              useQueryParams
                ? `params: removePathParams('${url}', data, true),`
                : `data: removePathParams('${url}', data, false),`
            }
            url: substituteParams('${url}', data),
          })
        }
      `;
    })
    .join('\n\n')}

  async paginate(request, data, config) {
    const result = [];

    let nextPageParams = {};
    do {
      const response = await this[request.name](
        { ...nextPageParams, ...data },
        config
      );

      result.push(...response.data.items);

      nextPageParams = response.data.links.next
        ? parseQueryParamsFromPagingLink(response.data.links.next)
        : {};
    } while (!!nextPageParams.nextPageToken);

    return result;
  }
}

module.exports.${outputClass} = ${outputClass};
  `.trim();

  return {
    javascript,
    declaration,
  };
};
