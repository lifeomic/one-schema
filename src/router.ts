import Router from '@koa/router';
import { z } from 'zod';
import { fromZodError } from 'zod-validation-error';
import zodToJsonSchema from 'zod-to-json-schema';
import compose = require('koa-compose');
import { IntrospectionConfig } from './koa';
import { EndpointImplementation, implementRoute } from './koa-utils';
import { IntrospectionResponse, OneSchemaDefinition } from './types';

export type RouterEndpointDefinition<Name> = {
  name: Name;
  description?: string;
  request: z.ZodType<any, any, any>;
  response: z.ZodType<any, any, any>;
};

type Method = 'GET' | 'DELETE' | 'PUT' | 'POST' | 'PATCH';
type RoughRoute = `${Method} ${string}`;

type ZodSchema = {
  [route: string]: RouterEndpointDefinition<string>;
};

export type OneSchemaRouterConfig<R extends Router<any, any>> = {
  using: R;
  introspection: IntrospectionConfig | undefined;
};

export class OneSchemaRouter<
  Schema extends ZodSchema,
  R extends Router<any, any>,
> {
  private router: R;

  private constructor(
    private schema: Schema,
    { introspection, using: router }: OneSchemaRouterConfig<R>,
  ) {
    this.router = router;
    if (introspection) {
      router.get(introspection.route, (ctx, next) => {
        const response: IntrospectionResponse = {
          serviceVersion: introspection.serviceVersion,
          schema: convertRouterSchemaToJSONSchemaStyle(this.schema),
        };

        ctx.body = response;
        ctx.status = 200;
        return next();
      });
    }
  }

  static create<R extends Router<any, any>>(
    config: OneSchemaRouterConfig<R>,
  ): OneSchemaRouter<{}, R> {
    return new OneSchemaRouter({}, config);
  }

  route<
    Route extends RoughRoute,
    Name extends string,
    Endpoint extends RouterEndpointDefinition<Name>,
  >(
    endpoint: Endpoint & { route: Route },
  ): OneSchemaRouter<Schema & { [route in Route]: Endpoint }, R> {
    // @ts-expect-error
    this.schema[endpoint.route] = endpoint;

    return this as any;
  }

  implement<Route extends keyof Schema & string>(
    route: Route,
    implementation: EndpointImplementation<
      Route,
      z.infer<Schema[Route]['request']>,
      z.infer<Schema[Route]['response']>,
      R
    >,
  ) {
    const endpoint = this.schema[route];

    implementRoute(
      route,
      this.router,
      (ctx, data) => {
        const res = endpoint.request.safeParse(data);
        if (!res.success) {
          const friendlyError = fromZodError(res.error, {
            prefix: 'The request input did not conform to the required schema',
          });
          return ctx.throw(400, friendlyError.message);
        }
        return res.data;
      },
      implementation,
    );

    return this;
  }

  middleware(): Router.Middleware {
    return compose([this.router.routes(), this.router.allowedMethods()]);
  }
}

const convertRouterSchemaToJSONSchemaStyle = <Schema extends ZodSchema>(
  schema: Schema,
): OneSchemaDefinition => {
  const oneSchema: OneSchemaDefinition = { Endpoints: {} };

  for (const [endpoint, definition] of Object.entries(schema)) {
    oneSchema.Endpoints[endpoint] = {
      Name: definition.name,
      Description: definition.description,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      Request: zodToJsonSchema(definition.request) as any,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      Response: zodToJsonSchema(definition.response) as any,
    };
  }

  return oneSchema;
};
