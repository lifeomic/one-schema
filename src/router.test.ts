import axios, { AxiosInstance } from 'axios';
import { Server } from 'http';
import { v4 as uuid } from 'uuid';
import { format } from 'prettier';
import Koa = require('koa');
import Router = require('@koa/router');
import bodyparser = require('koa-bodyparser');
import { OneSchemaRouter } from './router';
import { z } from 'zod';
import { generateAxiosClient } from './generate-axios-client';

let server: Server | undefined = undefined;
afterEach(() => {
  server?.close();
});

const setup = <T extends OneSchemaRouter<any, any>>(
  expose: (router: OneSchemaRouter<{}, Router>) => T,
): { client: AxiosInstance } => {
  const router = expose(
    OneSchemaRouter.create({ using: new Router(), introspection: undefined }),
  );
  return serve(router);
};

const serve = (
  router: OneSchemaRouter<{}, Router>,
): { client: AxiosInstance } => {
  server = new Koa().use(bodyparser()).use(router.middleware()).listen();

  const { port } = server.address() as any;

  const client = axios.create({
    baseURL: `http://localhost:${port}`,
    validateStatus: () => true,
  });

  return { client };
};

test('introspection', async () => {
  const { client } = setup(() =>
    OneSchemaRouter.create({
      using: new Router(),
      introspection: { route: '/private/introspection', serviceVersion: '123' },
    })
      .declare({
        name: 'getSomething',
        route: 'GET /something/:id',
        description: 'it gets something',
        request: z.object({ filter: z.string() }),
        response: z.object({ message: z.string(), id: z.string() }),
      })
      .implement('GET /something/:id', () => ({ id: '', message: '' }))
      .declare({
        name: 'createSomething',
        route: 'POST /something',
        description: 'it creates something',
        request: z.object({ message: z.string() }),
        response: z.object({ message: z.string(), id: z.string() }),
      })
      .implement('POST /something', () => ({ id: '', message: '' })),
  );

  const result = await client.get('/private/introspection');

  expect(result.status).toStrictEqual(200);
  expect(result.data).toStrictEqual({
    serviceVersion: '123',
    schema: {
      Endpoints: {
        'GET /something/:id': {
          Description: 'it gets something',
          Name: 'getSomething',
          Request: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              filter: { type: 'string' },
            },
            required: ['filter'],
            type: 'object',
          },
          Response: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              id: { type: 'string' },
              message: { type: 'string' },
            },
            required: ['message', 'id'],
            type: 'object',
          },
        },
        'POST /something': {
          Description: 'it creates something',
          Name: 'createSomething',
          Request: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              message: { type: 'string' },
            },
            required: ['message'],
            type: 'object',
          },
          Response: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              id: { type: 'string' },
              message: { type: 'string' },
            },
            required: ['message', 'id'],
            type: 'object',
          },
        },
      },
    },
  });
});

test('introspection with custom router', async () => {
  const { client } = setup(() =>
    OneSchemaRouter.create({
      using: new Router(),
      introspection: {
        route: '/private/introspection',
        serviceVersion: '123',
        router: new Router({ prefix: '/custom' }),
      },
    })
      .declare({
        name: 'getSomething',
        route: 'GET /something/:id',
        description: 'it gets something',
        request: z.object({ filter: z.string() }),
        response: z.object({ message: z.string(), id: z.string() }),
      })
      .implement('GET /something/:id', () => ({ id: '', message: '' })),
  );

  const wrongIntroRouter = await client.get('/private/introspection');
  expect(wrongIntroRouter.status).toStrictEqual(404);

  const result = await client.get('/custom/private/introspection');

  expect(result.data).toStrictEqual({
    serviceVersion: '123',
    schema: {
      Endpoints: {
        'GET /something/:id': {
          Description: 'it gets something',
          Name: 'getSomething',
          Request: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              filter: { type: 'string' },
            },
            required: ['filter'],
            type: 'object',
          },
          Response: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            additionalProperties: false,
            properties: {
              id: { type: 'string' },
              message: { type: 'string' },
            },
            required: ['message', 'id'],
            type: 'object',
          },
        },
      },
    },
  });
});

describe('type inference', () => {
  test('type inference for implementation return type', () => {
    setup((router) =>
      router
        .declare({
          name: 'getItem',
          route: 'GET /items/:id',
          request: z.object({ message: z.string() }),
          response: z.object({
            id: z.string(),
            message: z.string(),
          }),
        })
        .implement(
          'GET /items/:id',
          // This should not compile, since we're returning an object that does not match the schema.
          // @ts-expect-error
          () => ({ bogus: 'data' }),
        ),
    );
  });

  test('type inference for POST body + params', () => {
    setup((router) =>
      router
        .declare({
          name: 'createItem',
          route: 'POST /items/:id',
          request: z.object({ message: z.string() }),
          response: z.object({ message: z.string() }),
        })
        .implement('POST /items/:id', (ctx) => {
          // this statement helps us validate that the body is typed correctly
          ctx.request.body.message.trim();

          // This expect-error helps us confirm that the query is not on the ctx.
          // @ts-expect-error
          ctx.request.query;

          // this statement helps us validate that the params are typed correctly
          ctx.params.id.trim();

          return ctx.request.body;
        }),
    );
  });

  test('type inference for GET body + params', () => {
    setup((router) =>
      router
        .declare({
          name: 'get',
          route: 'GET /items/:id',
          request: z.object({ message: z.string() }),
          response: z.object({ message: z.string() }),
        })
        .implement('GET /items/:id', (ctx) => {
          // this statement helps us validate that the body is typed correctly
          ctx.request.query.message.trim();

          // This expect-error helps us confirm that the query is not on the ctx.
          // @ts-expect-error
          ctx.request.body;

          // this statement helps us validate that the params are typed correctly
          ctx.params.id.trim();

          return ctx.request.query;
        }),
    );
  });
});

describe('input validation', () => {
  (['POST', 'PUT', 'PATCH'] as const).forEach((method) => {
    test(`rejects requests that do not match the schema for ${method} requests`, async () => {
      const { client } = setup((router) =>
        router
          .declare({
            name: 'createItem',
            route: `${method} /items`,
            request: z.object({ message: z.string() }),
            response: z.object({ message: z.string() }),
          })
          .implement(`${method} /items`, () => ({ message: '' })),
      );

      const { status, data } = await client.request({
        method,
        url: '/items',
        data: { bogus: 'value' },
      });

      expect(status).toStrictEqual(400);
      expect(data).toStrictEqual(
        'The request input did not conform to the required schema: Required at "message"',
      );
    });
  });

  (['GET', 'DELETE'] as const).forEach((method) => {
    test(`rejects requests that do not match the schema for ${method} requests`, async () => {
      const { client } = setup((router) =>
        router
          .declare({
            name: 'getItem',
            route: `${method} /items`,
            request: z.object({ message: z.string() }),
            response: z.object({ message: z.string() }),
          })
          .implement(`${method} /items`, () => ({ message: '' })),
      );

      const { status, data } = await client.request({
        method,
        url: '/items',
        params: { bogus: 'value' },
      });

      expect(status).toStrictEqual(400);
      expect(data).toStrictEqual(
        'The request input did not conform to the required schema: Required at "message"',
      );
    });
  });

  test('error messages when there are multiple validation errors', async () => {
    const { client } = setup((router) =>
      router
        .declare({
          name: 'createItem',
          route: 'POST /items',
          request: z.object({ message: z.string(), private: z.boolean() }),
          response: z.object({ message: z.string() }),
        })
        .implement('POST /items', () => ({ message: '' })),
    );

    const { status, data } = await client.request({
      method: 'POST',
      url: '/items',
      data: { bogus: 'value' },
    });

    expect(status).toStrictEqual(400);
    expect(data).toStrictEqual(
      'The request input did not conform to the required schema: Required at "message"; Required at "private"',
    );
  });
});

describe('implementations', () => {
  (['POST', 'PUT', 'PATCH'] as const).forEach((method) => {
    test(`rejects requests that do not match the schema for ${method} requests`, async () => {
      const { client } = setup((router) =>
        router
          .declare({
            name: 'createItem',
            route: `${method} /items`,
            request: z.object({ message: z.string() }),
            response: z.object({ message: z.string() }),
          })
          .implement(`${method} /items`, (ctx) => ({
            message: ctx.request.body.message + '-response',
          })),
      );

      const message = uuid();

      const { status, data } = await client.request({
        method,
        url: '/items',
        data: { message },
      });

      expect(status).toStrictEqual(200);
      expect(data).toStrictEqual({ message: `${message}-response` });
    });
  });

  (['GET', 'DELETE'] as const).forEach((method) => {
    test(`rejects requests that do not match the schema for ${method} requests`, async () => {
      const { client } = setup((router) =>
        router
          .declare({
            name: 'createItem',
            route: `${method} /items`,
            request: z.object({ message: z.string() }),
            response: z.object({ message: z.string() }),
          })
          .implement(`${method} /items`, (ctx) => ({
            message: (ctx.request.query.message as string) + '-response',
          })),
      );

      const message = uuid();

      const { status, data } = await client.request({
        method,
        url: '/items',
        params: { message },
      });

      expect(status).toStrictEqual(200);
      expect(data).toStrictEqual({ message: `${message}-response` });
    });
  });
});

describe('introspection', () => {
  test('introspecting + generating a client', async () => {
    const { client } = serve(
      OneSchemaRouter.create({
        using: new Router(),
        introspection: {
          route: '/private/introspection',
          serviceVersion: '123',
        },
      })
        .declare({
          name: 'createItem',
          route: 'POST /items',
          request: z.object({ message: z.string() }),
          response: z.object({ id: z.string(), message: z.string() }),
        })
        .implement('POST /items', (ctx) => ({
          id: 'something',
          message: ctx.request.body.message + '-response',
        }))
        .declare({
          route: 'GET /items/:id',
          name: 'getItem',
          request: z.object({
            filter: z.string(),
          }),
          response: z.object({ id: z.string(), message: z.string() }),
        })
        .implement('GET /items/:id', (ctx) => ({
          id: ctx.params.id,
          message: ctx.request.query.filter + '-response',
        })),
    );

    const { status, data } = await client.request({
      method: 'GET',
      url: '/private/introspection',
    });

    expect(status).toStrictEqual(200);
    expect(data).toStrictEqual({
      serviceVersion: '123',
      schema: {
        Endpoints: expect.objectContaining({
          'POST /items': expect.anything(),
          'GET /items/:id': expect.anything(),
        }),
      },
    });

    const clientCode = await generateAxiosClient({
      spec: data.schema,
      outputClass: 'Client',
    });

    const formattedDeclaration = format(clientCode.declaration, {
      parser: 'typescript',
    });

    expect(formattedDeclaration).toMatchSnapshot();
  });

  test('when using the same schema reference multiple times, it is always resolved inline', async () => {
    const reusedSchema = z.object({ id: z.string() });

    const router = OneSchemaRouter.create({
      using: new Router(),
      introspection: {
        route: '/private/introspection',
        serviceVersion: '123',
      },
    }).declare({
      route: 'POST /items',
      name: 'createItem',
      request: z.object({}),
      response: z.object({
        resProp1: reusedSchema,
        resProp2: reusedSchema,
      }),
    });

    const { client } = serve(router);

    const introspectionResult = await client.get('/private/introspection');

    console.log(JSON.stringify(introspectionResult.data, null, 2));

    expect(introspectionResult.data.schema).toStrictEqual({
      Endpoints: {
        'POST /items': {
          Name: 'createItem',
          Request: {
            type: 'object',
            properties: {},
            additionalProperties: false,
            $schema: 'http://json-schema.org/draft-07/schema#',
          },
          Response: {
            type: 'object',
            properties: {
              // resProp1 and resProp2 should be inlined, even though they have identical
              // schemas defined using the same reference.
              resProp1: {
                type: 'object',
                properties: {
                  id: {
                    type: 'string',
                  },
                },
                required: ['id'],
                additionalProperties: false,
              },
              resProp2: {
                type: 'object',
                properties: {
                  id: {
                    type: 'string',
                  },
                },
                required: ['id'],
                additionalProperties: false,
              },
            },
            required: ['resProp1', 'resProp2'],
            additionalProperties: false,
            $schema: 'http://json-schema.org/draft-07/schema#',
          },
        },
      },
    });
  });
});

test('declaring multiple routes with the same name results in an error', () => {
  expect(() => {
    OneSchemaRouter.create({
      using: new Router(),
      introspection: undefined,
    })
      .declare({
        route: 'POST /items',
        name: 'createSomething',
        request: z.object({}),
        response: z.object({}),
      })
      .declare({
        route: 'PUT /items',
        name: 'createSomething',
        request: z.object({}),
        response: z.object({}),
      });
  }).toThrowError(
    'Multiple endpoints were declared with the same name "createSomething". Each endpoint must have a unique name.',
  );
});

test('the client(...) helper', async () => {
  const router = OneSchemaRouter.create({
    using: new Router(),
    introspection: undefined,
  })
    .declare({
      route: 'POST /items',
      name: 'createItem',
      request: z.object({ id: z.string() }),
      response: z.object({ id: z.string() }),
    })
    .implement('POST /items', (ctx) => ({ id: ctx.request.body.id }))
    .declare({
      route: 'GET /items/:id',
      name: 'getItemById',
      request: z.object({
        filter: z.string(),
      }),
      response: z.object({ id: z.string() }),
    })
    .implement('GET /items/:id', (ctx) => ({
      id: ctx.params.id + ':' + ctx.request.query.filter,
    }));

  const { client: axios } = serve(router);

  const client = router.client(axios);

  const getResponse = await client.getItemById({
    id: 'some-id',
    filter: 'some-filter',
  });

  expect(getResponse.status).toStrictEqual(200);
  expect(getResponse.data).toStrictEqual({ id: 'some-id:some-filter' });

  const postResponse = await client.createItem({ id: 'some-id' });
  expect(postResponse.status).toStrictEqual(200);
  expect(postResponse.data).toStrictEqual({ id: 'some-id' });
});
