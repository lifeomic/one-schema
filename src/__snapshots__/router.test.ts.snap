// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspection introspecting + generating a client 1`] = `
"/* eslint-disable */
import { AxiosInstance, AxiosRequestConfig, AxiosResponse } from \\"axios\\";

export type Endpoints = {
  \\"POST /items\\": {
    Request: {
      message: string;
    };
    PathParams: {};
    Response: {
      id: string;
      message: string;
    };
  };
  \\"GET /items/:id\\": {
    Request: {
      filter: string;
    };
    PathParams: {
      id: string;
    };
    Response: {
      id: string;
      message: string;
    };
  };
};

export declare class Client {
  constructor(client: AxiosInstance);

  /**
   * Executes the \`POST /items\` endpoint.
   *
   * @param data The request data.
   * @param config The Axios request overrides for the request.
   *
   * @returns An AxiosResponse object representing the response.
   */
  createItem(
    data: Endpoints[\\"POST /items\\"][\\"Request\\"] &
      Endpoints[\\"POST /items\\"][\\"PathParams\\"],
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<Endpoints[\\"POST /items\\"][\\"Response\\"]>>;

  /**
   * Executes the \`GET /items/:id\` endpoint.
   *
   * @param data The request data.
   * @param config The Axios request overrides for the request.
   *
   * @returns An AxiosResponse object representing the response.
   */
  getItem(
    data: Endpoints[\\"GET /items/:id\\"][\\"Request\\"] &
      Endpoints[\\"GET /items/:id\\"][\\"PathParams\\"],
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<Endpoints[\\"GET /items/:id\\"][\\"Response\\"]>>;

  /**
   * Paginates exhaustively through the provided \`request\`, using the specified
   * \`data\`. A \`pageSize\` can be specified in the \`data\` to customize the
   * page size for pagination.
   */
  paginate<T extends { nextPageToken?: string; pageSize?: string }, Item>(
    request: (
      data: T,
      config?: AxiosRequestConfig
    ) => Promise<
      AxiosResponse<{ items: Item[]; links: { self: string; next?: string } }>
    >,
    data: T,
    config?: AxiosRequestConfig
  ): Promise<Item[]>;
}
"
`;
